lib/insulet/insuletSimulator.js

```diff
@@ -0,0 +1,376 @@
+  }
+
+  function setCurrBolus(bolus) {
+    currBolus = bolus;
+  }
+
+  function setCurrStatus(status) {
+    currStatus = status;
+  }
+
+  function ensureTimestamp(e){
+    if (currTimestamp > e.time) {
+      throw new Error(
+        util.format('Timestamps must be in order.  Current timestamp was[%s], but got[%j]', currTimestamp, e)
+      );
+    }
+    currTimestamp = e.time;
+    return e;
+  }
+
+  function simpleSimulate(e) {
+    ensureTimestamp(e);
+    events.push(e);
+  }
+
+  function fillInSuppressed(e) {
+    var schedName = null, suppressed = _.clone(e.suppressed);
+    if (e.previous != null) {
+      if (e.previous.deliveryType === 'scheduled') {
+        schedName = e.previous.scheduleName;
+      }
+      else if (e.previous.deliveryType === 'temp') {
+        if (e.previous.suppressed != null) {
+          if (e.previous.suppressed.deliveryType === 'scheduled') {
+            schedName = e.previous.suppressed.scheduleName;
+          }
+        }
+      }
+    }
+
+    if (schedName != null) {
+      e.suppressed = suppressed.with_scheduleName(schedName).done();
+    }
+    else {
+      delete e.suppressed;
+    }
+  }
+
+  return {
+    alarm: function(event) {
+      if (event.payload != null && event.payload.stopsDelivery === true) {
+        if (event.status == null) {
+          throw new Error('An Insulet alarm that has `stopsDelivery` in the payload must have a `status`.');
+        }
+      }
+      simpleSimulate(event);
+    },
+    basal: function(event){
+      ensureTimestamp(event);
+      if (currBasal != null) {
+        // sometimes there can be duplicate suspend basals, so we return early
+        // if we come across a suspend basal when we're already in one
+        if (currBasal.deliveryType === 'suspend' && event.deliveryType === 'suspend') {
+          return;
+        }
+        // completing a resume event from a new pod activation
+        // see podActivation() below for more details
+        if (currStatus != null && currStatus.status === 'suspended') {
+          if (activationStatus != null && activationStatus.status === 'resumed') {
+            var resume = activationStatus.with_previous(_.omit(currStatus, 'previous'))
+              .with_reason({resumed: 'manual'})
+              .done();
+            setCurrStatus(resume);
+            events.push(resume);
+            setActivationStatus(null);
+          }
+          else {
+            // if we're in a suspend basal, we need to leave the currStatus because
+            // a resume is probably still impending
+            // but if we've already moved on to a basal that's not deliveryType: suspend
+            // then we need to wipe the slate clean
+            if (event.deliveryType !== 'suspend') {
+              setCurrStatus(null);
+            }
+          }
+        }
+        // currBasal.duration is a string when the builder object hasn't had
+        // .with_duration called yet (i.e., it's the string `**REQUIRED**`)
```

Comment:
Should we perhaps add a helper function to objectBuilder that would let you call something like isAssigned('duration'), rather than depending on the implementation details of objectBuilder?